boot/systems.go

```diff
@@ -342,3 +343,84 @@ func InspectTryRecoverySystemOutcome(dev Device) (outcome TryRecoverySystemOutco
 
 	return outcome, trySystem, nil
 }
+
+// PromoteTriedRecoverySystem promotes the provided recovery system to be
+// recognized as a good one, and ensures that the system is present in the list
+// of good recovery systems and current recovery systems in modeenv. The
+// provided list of tried systems should contain the system in question. If the
+// system uses encryption, the keys will updated state. If resealing fails, an
+// attempt to restore the previous state is made
+func PromoteTriedRecoverySystem(dev Device, systemLabel string, triedSystems []string) (err error) {
+	if !dev.HasModeenv() {
+		return fmt.Errorf("internal error: recovery systems can only be used on UC20")
+	}
+
+	if !strutil.ListContains(triedSystems, systemLabel) {
+		// system is not among the tried systems
+		return fmt.Errorf("system has not been successfully tried")
+	}
+
+	m, err := loadModeenv()
+	if err != nil {
+		return err
+	}
+	rewriteModeenv := false
+	if !strutil.ListContains(m.CurrentRecoverySystems, systemLabel) {
+		m.CurrentRecoverySystems = append(m.CurrentRecoverySystems, systemLabel)
+		rewriteModeenv = true
+	}
+	if !strutil.ListContains(m.GoodRecoverySystems, systemLabel) {
+		m.GoodRecoverySystems = append(m.GoodRecoverySystems, systemLabel)
+		rewriteModeenv = true
+	}
+	if rewriteModeenv {
+		if err := m.Write(); err != nil {
+			return err
+		}
+	}
+
+	defer func() {
```

Comment:
(nitpick) Given that the defer is added here and the next thing is the `return resealKeyToModeenv(dirs.GlobalRootDir, dev.Model(), m, expectReseal)` - would it make sense to simply write this as:
```
	const expectReseal = true
        if err := resealKeyToModeenv(dirs.GlobalRootDir, dev.Model(), m, expectReseal); err != nil {
            if cleanupErr := DropRecoverySystem(dev, systemLabel); cleanupErr != nil {
    			err = fmt.Errorf("%v (cleanup failed: %v)", err, cleanupErr)
            }
            return err
       }
       return nil
}
```
i.e. just avoid the slightly harder to read defer?

Or will this code change in the future and then the defer will be used for more than this one  failure case of "resealKeyToModeenv" ?

Reply:
Thanks for the suggestion, it looks nicer indeed.